### ISSUE TITLE
CMake cleans all MPI flags before build

### ISSUE THREAD

**Issue by njzjz** [merzlab/QUICK#343]
Early this year, I worked to add an MPI build for AmberTools in conda-forge (https://github.com/conda-forge/ambertools-feedstock/pull/133). I got the error `mpi.h: No such file or directory` when compiling QUICK using MPI.

After careful debugging, I found that the following CMake lines clean all compiling flags, including the path to the MPI directory (something like `-I /path/to/mpi/include`).

https://github.com/merzlab/QUICK/blob/fd1ccc4ec0f59ff63f9f74cb895b38c4e3df68b6/CMakeLists.txt#L58-L60

I added a patch to remove the above lines (https://github.com/conda-forge/ambertools-feedstock/pull/133/commits/b9dc271539941373a0b2a54c3e3a70868f2f261e), and now it works fine.

I am unsure whether this patch can be ported to the upstream source code here, as I am unsure what these lines are used for.

**Comment by agoetz** [comment]
I never encountered such an error when compiling AmberTools. I am not sure what went wrong for you on conda-forge.

Because QUICK is a standalone code we generate our own compiler flags. In general we do not want to use flags set for other codes in AmberTools. That said, integration with AmberTools can probably be improved.

@ohearnk modernized the MPI implementation in QUICK, we now use the Fortran mpi module, see #328 which was merged into QUICK master 3 days ago. This will go into AmberTools24 and will be merged over the next few days.

@ohearnk plase clarify if I misunderstand something. I think we can close this PR, if yes, please do so. Thanks.

**Comment by ohearnk** [comment]
@agoetz I believe this is fixed as well.  Just to make sure, let's hold off closing this until the first AmberTools 2024 RC comes out.  We'll verify then that this is resolved.

**Comment by njzjz** [comment]
The error still exists in AmberTools 24-rc2:

```
[ 36%] Built target sander_rism_interface_mpi
[ 36%] Building CXX object AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/octree.cpp.o                                                   In file included from /home/conda/feedstock_root/build_artifacts/ambertools_1712111999571/work/AmberTools/src/quick/src/octree/octree.cpp:18:
/home/conda/feedstock_root/build_artifacts/ambertools_1712111999571/work/AmberTools/src/quick/src/octree/octree.h:8:10: fatal error: mpi.h: No such
file or directory
    8 | #include <mpi.h>
      |          ^~~~~~~
compilation terminated.
make[2]: *** [AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/build.make:76: AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/octree.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:6643: AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/all] Error 2                                     make: *** [Makefile:156: all] Error 2
```

**Comment by agoetz** [comment]
OK, this is a single include in C++ code, which must remain.

I never used conda-forge and do not know how it works. How does AmberTools get configured on conda-forge? What is different from me executing cmake followed by make on the command line?

@ohearnk can you take a look?

**Comment by njzjz** [comment]
A detailed log using `make VERBOSE=1`:

```
[ 36%] Built target sander_rism_interface_mpi
make  -f AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/build.make AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/depend
make[2]: Entering directory '$SRC_DIR/build'
cd $SRC_DIR/build && $BUILD_PREFIX/bin/cmake -E cmake_depends "Unix Makefiles" $SRC_DIR $SRC_DIR/AmberTools/src/quick/src/octree $SRC_DIR/build $SRC_DIR/build/AmberTools/src/quick/src/octree $SRC_DIR/build/AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/DependInfo.cmake "--color="
make[2]: Leaving directory '$SRC_DIR/build'
make  -f AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/build.make AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/build
make[2]: Entering directory '$SRC_DIR/build'
[ 36%] Building CXX object AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/octree.cpp.o
cd $SRC_DIR/build/AmberTools/src/quick/src/octree && $BUILD_PREFIX/bin/x86_64-conda-linux-gnu-c++ -DCEW -DGNU -DMPIV -I$SRC_DIR/AmberTools/src/quick/src/octree/. -std=gnu++11 -fPIC -O2 -mtune=native -pthread -MD -MT AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/octree.cpp.o -MF CMakeFiles/octree_mpi.dir/octree.cpp.o.d -o CMakeFiles/octree_mpi.dir/octree.cpp.o -c $SRC_DIR/AmberTools/src/quick/src/octree/octree.cpp
In file included from /home/conda/feedstock_root/build_artifacts/ambertools_1712113373389/work/AmberTools/src/quick/src/octree/octree.cpp:18:
/home/conda/feedstock_root/build_artifacts/ambertools_1712113373389/work/AmberTools/src/quick/src/octree/octree.h:8:10: fatal error: mpi.h: No such file or directory
8 | #include <mpi.h>
|          ^~~~~~~
compilation terminated.
make[2]: *** [AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/build.make:76: AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/octree.cpp.o] Error 1
make[2]: Leaving directory '$SRC_DIR/build'
make[1]: *** [CMakeFiles/Makefile2:6643: AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/all] Error 2
make[1]: Leaving directory '$SRC_DIR/build'
make: *** [Makefile:156: all] Error 2    
```

[build.log](https://github.com/merzlab/QUICK/files/14846668/build.log)

**Comment by agoetz** [comment]
Something seems missing in your environment. I see this in my Linux box
```
cd build
touch ../AmberTools/src/quick/src/octree/octree.cpp
make VERBOSE=TRUE quick.MPI
 -- snip --
make  -f AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/build.make AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/depend
make[3]: Entering directory `/scratch/agoetz/amber24-rc2/amber24_src/build'
cd /scratch/agoetz/amber24-rc2/amber24_src/build && /netbin/CMake/cmake-3.28.1-linux-x86_64/bin/cmake -E cmake_depends "Unix Makefiles" /scratch/agoetz/amber24-rc2/amber24_src /scratch/agoetz/amber24-rc2/amber24_src/AmberTools/src/quick/src/octree /scratch/agoetz/amber24-rc2/amber24_src/build /scratch/agoetz/amber24-rc2/amber24_src/build/AmberTools/src/quick/src/octree /scratch/agoetz/amber24-rc2/amber24_src/build/AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/DependInfo.cmake "--color="
Dependencies file "AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/octree.cpp.o.d" is newer than depends file "/scratch/agoetz/amber24-rc2/amber24_src/build/AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/compiler_depend.internal".
Consolidate compiler generated dependencies of target octree_mpi
make[3]: Leaving directory `/scratch/agoetz/amber24-rc2/amber24_src/build'
make  -f AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/build.make AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/build
make[3]: Entering directory `/scratch/agoetz/amber24-rc2/amber24_src/build'
[ 84%] Building CXX object AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/octree.cpp.o
cd /scratch/agoetz/amber24-rc2/amber24_src/build/AmberTools/src/quick/src/octree && /opt/rh/devtoolset-10/root/usr/bin/g++ -DCEW -DGNU -DMPIV -I/scratch/agoetz/amber24-rc2/amber24_src/AmberTools/src/quick/src/octree/. -I/netbin/mpi/openmpi-4.0.5-gnu-8.3.1/include -std=gnu++11 -fPIC -O2 -mtune=native -pthread -MD -MT AmberTools/src/quick/src/octree/CMakeFiles/octree_mpi.dir/octree.cpp.o -MF CMakeFiles/octree_mpi.dir/octree.cpp.o.d -o CMakeFiles/octree_mpi.dir/octree.cpp.o -c /scratch/agoetz/amber24-rc2/amber24_src/AmberTools/src/quick/src/octree/octree.cpp
[ 84%] Linking CXX static library liboctree_mpi.a
```
As you can see, it picks up the include directory `-I/netbin/mpi/openmpi-4.0.5-gnu-8.3.1/include`. 

I have this in my module file
```
[agoetz@chinotto build]$ module show openmpi/4.0.5-gnu-8.3.1
-------------------------------------------------------------------
/netbin/modulefiles/openmpi/4.0.5-gnu-8.3.1:

conflict	 mpich openmpi 
setenv		 MPI_BIN /netbin/mpi/openmpi-4.0.5-gnu-8.3.1/bin 
setenv		 MPI_INCLUDE /netbin/mpi/openmpi-4.0.5-gnu-8.3.1/include 
setenv		 MPI_LIB /netbin/mpi/openmpi-4.0.5-gnu-8.3.1/lib 
setenv		 MPI_MAN /netbin/mpi/openmpi-4.0.5-gnu-8.3.1/share/man 
setenv		 MPI_COMPILER openmpi-x86_64 
setenv		 MPI_SUFFIX _openmpi 
setenv		 MPI_HOME /netbin/mpi/openmpi-4.0.5-gnu-8.3.1 
prepend-path	 PATH /netbin/mpi/openmpi-4.0.5-gnu-8.3.1/bin 
prepend-path	 LD_LIBRARY_PATH /netbin/mpi/openmpi-4.0.5-gnu-8.3.1/lib 
prepend-path	 MANPATH :/netbin/mpi/openmpi-4.0.5-gnu-8.3.1/share/man 
-------------------------------------------------------------------
```
I used this cmake command 
```
#  Assume this is Linux:

  cmake $AMBER_PREFIX/amber24_src \
    -DCMAKE_INSTALL_PREFIX=$AMBER_PREFIX/amber24 \
    -DCOMPILER=GNU  \
    -DMPI=TRUE -DCUDA=TRUE -DINSTALL_TESTS=TRUE \
    -DDOWNLOAD_MINICONDA=TRUE \
    2>&1 | tee  cmake.log
```

**Comment by njzjz** [comment]
One thing is different in conda-forge: conda-forge uses `-DTRUST_SYSTEM_LIBS=TRUE`.

**Comment by hanahg** [comment]
Hi,
I have the same problem. please let me know if the problem is solved for you and how?

**Comment by agoetz** [comment]
Hi @hanahg. Can you please give more details? Are you compiling QUICK 24.03 release or the QUICK master branch? Are you trying to compile AmberTools 24? How does your cmake command look like?

**Comment by hanahg** [comment]
Hi
I am installing AMBER24&AmberTools24 along with.

  cmake $AMBER_PREFIX/amber24_src \
    -DCMAKE_INSTALL_PREFIX=$AMBER_PREFIX/amber24 \
    -DCOMPILER=AUTO  \
    -DMPI=FALSE -DCUDA=FALSE -DINSTALL_TESTS=TRUE \
    -DDOWNLOAD_MINICONDA=FALSE \
    2>&1 | tee  cmake.log

I have changed -DMPI =TRUE, -DINSTALL_TESTS=FALSE also. I have cuda 12.6 which is not compatible with. Amber installed with no error but the test file showed errors.

Thank you for your help

Envoyé à partir de Outlook<http://aka.ms/weboutlook>
________________________________

**Comment by agoetz** [comment]
Hi @hanahg this does not seem to be related to this issue (CMake MPI flags in QUICK).

Please post your question to the Amber mailing list. When posting, please describe clearly which commands you use and what goes wrong. Include relevant error messages. Otherwise there is no way to help.

If it turns out that you encountered an issue with QUICK 24.03 in AmberTools 24, then we can follow up either on the Amber mailing list or you can open a discussion on the QUICK Github repo (please do not open an issue unless it is confirmed there that there is an issue with the source code). Thanks.
